pkg/devfile/generator/generators.go

```diff
@@ -0,0 +1,440 @@
+		envVars := convertEnvs(comp.Container.Env)
+		resourceReqs := getResourceReqs(comp)
+		ports := convertPorts(comp.Container.Endpoints)
+		containerParams := containerParams{
+			Name:         comp.Name,
+			Image:        comp.Container.Image,
+			IsPrivileged: false,
+			Command:      comp.Container.Command,
+			Args:         comp.Container.Args,
+			EnvVars:      envVars,
+			ResourceReqs: resourceReqs,
+			Ports:        ports,
+		}
+		container := getContainer(containerParams)
+
+		// If `mountSources: true` was set PROJECTS_ROOT & PROJECT_SOURCE env
+		if comp.Container.MountSources == nil || *comp.Container.MountSources {
+			syncRootFolder := addSyncRootFolder(container, comp.Container.SourceMapping)
+
+			err := addSyncFolder(container, syncRootFolder, devfileObj.Data.GetProjects())
+			if err != nil {
+				return nil, err
+			}
+		}
+		containers = append(containers, *container)
+	}
+	return containers, nil
+}
+
+// PodTemplateSpecParams is a struct that contains the required data to create a pod template spec object
+type PodTemplateSpecParams struct {
+	ObjectMeta     metav1.ObjectMeta
+	InitContainers []corev1.Container
+	Containers     []corev1.Container
+	Volumes        []corev1.Volume
+}
+
+// GetPodTemplateSpec gets a pod template spec that can be used to create a deployment spec
+func GetPodTemplateSpec(podTemplateSpecParams PodTemplateSpecParams) *corev1.PodTemplateSpec {
+	podTemplateSpec := &corev1.PodTemplateSpec{
+		ObjectMeta: podTemplateSpecParams.ObjectMeta,
+		Spec: corev1.PodSpec{
+			InitContainers: podTemplateSpecParams.InitContainers,
+			Containers:     podTemplateSpecParams.Containers,
+			Volumes:        podTemplateSpecParams.Volumes,
+		},
+	}
+
+	return podTemplateSpec
+}
+
+// DeploymentSpecParams is a struct that contains the required data to create a deployment spec object
+type DeploymentSpecParams struct {
+	PodTemplateSpec   corev1.PodTemplateSpec
+	PodSelectorLabels map[string]string
+}
+
+// GetDeploymentSpec gets a deployment spec
+func GetDeploymentSpec(deploySpecParams DeploymentSpecParams) *appsv1.DeploymentSpec {
+	deploymentSpec := &appsv1.DeploymentSpec{
+		Strategy: appsv1.DeploymentStrategy{
+			Type: appsv1.RecreateDeploymentStrategyType,
+		},
+		Selector: &metav1.LabelSelector{
+			MatchLabels: deploySpecParams.PodSelectorLabels,
+		},
+		Template: deploySpecParams.PodTemplateSpec,
+	}
+
+	return deploymentSpec
+}
+
+// DeploymentParams is a struct that contains the required data to create a deployment object
+type DeploymentParams struct {
+	TypeMeta       metav1.TypeMeta
+	ObjectMeta     metav1.ObjectMeta
+	DeploymentSpec appsv1.DeploymentSpec
+}
+
+// GetDeployment gets a deployment object
+func GetDeployment(deployParams DeploymentParams) *appsv1.Deployment {
+
+	deployment := &appsv1.Deployment{
+		TypeMeta:   deployParams.TypeMeta,
+		ObjectMeta: deployParams.ObjectMeta,
+		Spec:       deployParams.DeploymentSpec,
+	}
+
+	return deployment
```

Comment:
this is getting confusing, can we have a single function instead?
```
GetDeployment
GetDeploymentSpec
```

Reply:
Currently odo uses `GetDeploymentSpec` and wraps the spec with `deployment` when odo goes to create it. There is 3-4 lines redundant code in odo for every kube resources. How do you feel about `GetDeployment()` being the exposed func and it takes inputs for all the spec generation?

This would mean, odo would need to be modified a bit more.

Reply:
For example, like this https://github.com/openshift/odo/blob/master/pkg/kclient/deployments.go#L169

Reply:
Updated to have GetDeployment, GetService, GetIngress, etc as exposed func